### ISSUE TITLE
Hide FoldRegion end line

### ISSUE THREAD

**Issue by aminroosta** [microsoft/vscode#111731]
As discussed in #3352 :

**Comment by aeschli** [comment]
Yes, each folding range provider defines how the folding ranges exactly look like. It really depends on the language and the language's formatting conventions. That's why I why in the end each provider needs to come up with their own settings to tune the folding ranges, and why the users that want different folding behavior for a language should file it for that language. I tried to explain that in https://github.com/microsoft/vscode/issues/3352#issuecomment-513696419

A new setting like you suggest could only be supported for the indentation based folding, which we control. 
Maybe it's not even part of the indentation based folding, but a new kind of folding strategy, solely based on regexes.
Users can already define the folding strategy they want for a language: `editor.foldingStrategy`. But beware that it's very difficult to use regexes to parse a language.

**Comment by aeschli** [comment]
It's really a hack that you are suggesting, working only for a few cases where there is an end token. But in many cases the next folding range already starts on the next line. Just a example that comes to my mind:
```
switch (e) {
   case 1:
       //do this
       break:
   case 2:
      // do that
     break
}
```
You can experiment with that by implementing your own folding provider, copying the indentation based strategy and modifying it.
That said, we are lacking a solution when there are more than one providers. We would need a way for the user to select the preferred provider.

**Comment by aeschli** [comment]
May I suggest you continue to experiment by implementing your own folding provider? I'm not opposed to improve the indentation base formatting by additional setting(s), or have a different type of folding strategy based on regexes only (#104667)

But the feature must work for all type of syntaxes. Note that folding ranges can not overlap. You want to prove the soundness of the feature with tests against code examples of various languages.

Closing for now, happy to reopen or continue the discussion.

**Comment by aminroosta** [comment]
@aeschli I know realize the proper way to fix this is for folding providers to implement a compact mode.
Folding providers are per language, so it needs to be done for every language.

What i was initially suggesting was a hack, so lets ignore that all together.

In  #111921 I've tried to implement it for typescript/javascript.

**Comment by aminroosta** [comment]
The general idea is "Implement compact fold mode for all built in extensions".
If you agree, please feel free to leave comments on #111921 to clean it up (naming, unit tests (?), etc ...).

I'd be happy to send a PR for each built in extension over time.

**Comment by aminroosta** [comment]
@aeschli 
I was curious how much work it is to support it for another language, so i tried html as well.

PR:  https://github.com/microsoft/vscode-html-languageservice/pull/94 I've tried to implement in for html.
(needs a corresponding PR for the vscode repo).

Feel free to dismiss these if it doesn't make sense, it's totally fine.

**Comment by aeschli** [comment]
Thanks a lot @aminroosta, That's the right way!
But I wouldn't want to add a global 'editor.fold.compact' as it would be confusing for all languages that don't have such a mode.
Instead it should be each extension providing the setting.

**Comment by aminroosta** [comment]
Agreed, I actually didn’t add a global flag in the #111921.

**Comment by aminroosta** [comment]
@aeschli could you give an update on this?
I don't wanna close this again, like what happened to #9605

**Comment by aeschli** [comment]
Each PR will be looked at independently. I will look at the HTML proposal when I find time, @mjbvz the one for TypeScript.

**Comment by aeschli** [comment]
As discussed, this needs to be implemented on a per-language basis.
For HTML, the PR is https://github.com/microsoft/vscode-html-languageservice/pull/94
